### ISSUE TITLE
Replace Globalization Block [SystemJS Spike]

### ISSUE THREAD

**Issue by dchambers** [BladeRunnerJS/brjs#1416]
As a consequence of the [SystemJS spike](https://github.com/BladeRunnerJS/brjs/issues/1389) the dashboard is now running using SystemJS instead of browser-modules, but with some caveats due to globalization:

  1. I've so far been unable to stop ordering the bundle-set since we currently globalize thirdparty modules in-line, rather than waiting for all modules to be registered first.
  1. It's clear to me that we are going to need something like `globalizeSourceModules()` &mdash; even for future apps that don't use any namespaced code &mdash; so that we can force all modules to be reloaded synchronously within tests.
  1. From speaking to the mobile team, we can see that it would be beneficial if Bluebird was an optional dependency (promises are always available natively on mobile browsers and not having Bluebird would allow us to reduce the download by 11KB).
  1. Making Bluebird optional will require us to ensure there are no uses of `System.syncImport()` anywhere in this case.

## Optional Bluebird Support

So that Bluebird can be an optional dependency, it makes sense to create a `syncImport` module that depends on Bluebird rather than adding a `syncImport()` method to SystemJS, for example:

```js
var myModule = require('syncImport').import('some-module');
```

Now, given that SystemJS already does the right thing synchronously for bundled modules, and is only problematic for code external to the bundle, it would make sense if instead of a globalization block we had a `br/all` module that depended on all bundled modules, and on which you could invoke `globalize()`, so that we could write code like this:

```js
require('syncImport').import('br/all').globalize();
```

or for apps that don't need globalization, then we might instead write code like this:

```js
require('syncImport').import('br/all');
```

## Test Support

So that we can continue supporting the `globalizeSourceModules()` function 
function within tests, we could add an `install()` method to `syncImport` that makes the  `globalizeSourceModules()` function available globally, and then append code like this to the end of every test bundle:

```js
require('syncImport').import('br/all').globalize().install();
```

## Aspect / Workbench Support

Aspects and workbenches whose `index.html` does not invoke `System.import()`, like this for example:

```js
System.import('novox/App', function(module) {
	var App = module.default;
	var app = new App();
	app.start();
});
```

could automatically have the `require('syncImport').import('br/all').globalize()` line appended to the bundle, allowing existing aspect and workbench pages to continue working as before.

**Comment by dchambers** [comment]
@andyberry88, I've updated this proposal. Might be worth casting your eye over it at some point?

**Comment by dchambers** [comment]
I've now been able to verify that I can run JavaScript tests using the [system-js spike](https://github.com/BladeRunnerJS/brjs/issues/1389) provided those tests don't make use of the sub-realms feature.

Next, the following stack needs to be created:

  1. Create a 'syncImport' library on Github/NPM.
  1. Create a MockableSystemJS library on Github/NPM (will depend on 'syncImport').
  1. Create a 'sub-realms-compat' library on Github/NPM (will depend on 'MockableSystemJS').

and we'll then be able to complete these tasks:

  1. Replace the globalization block with `br/all` module (depends on 'syncImport').
  1. Ensure there is only a single use of `syncImport()` in the entire code-base (the one added automatically to test bundles and aspect/workbench bundles that don't do their own `System.import()`).
  1. Get the JavaScript tests working in my 'systemjs-spike' branch by using the sub-realms compatibilty library (depends on 'sub-realms-compat').

**Comment by dchambers** [comment]
From talking to @andyberry88 we wondered whether we may no longer need a globalizeSourceModules() method since:

  1. Installing a sub-realm will require us to invoke `require('syncImport').import('br/all')` even for code-bases that don't use namespaced code.
  1. For the automatic `require('syncImport').import('br/all')` that is added to the end of the bundle than it makes sense that we only invoke `globalize()` if there are modules that depend on globalization.

This will however require that 'sub-realms-compat' is a BRJS library since `br/all` is a BRJS concept, and since knowing whether or not to invoke `globalize()` on that module will require some tie-in with BRJS.

**Comment by dchambers** [comment]
My first attempt at replacing the globalization block with a `br/all` module didn't quite work because the alias and service modules attempt to export a service that is retrieved through an alias or service registry that doesn't depend on the necessary classes to use `require()`, and where `System.get()` only happens to work if the load order was such that the concrete dependencies happen to have been created at that point.

I verified that I could change the model to return `module` instead of `module.exports`, and only export the service after all the modules had been loaded, but this meant that people requiring a service now had to write code like this (breaking backwards compatibility):

```js
var myService = require('service!my-service').exports;
```

Realistically, there are only two practical options around this:

  1. Make the service registry dependent on every service (this would make it a special class with extra injected dependencies)?
  1. Get rid of the service registry?

**Comment by dchambers** [comment]
I just took a look at the [New Registry Proposal](https://github.com/BladeRunnerJS/brjs/wiki/New-Registry-Proposal#registry-proposal) to see if its features could be replicated in a system using only modules, and having no registries at all, and it seems do-able:

  * Service aquisition (`require('service!my-service')` or `System.get('service!my-service')`)
  * Fetch all services (`System.import('br/services').then(...)`)
  * Fetch named services (`System.import()` + `Promise.all()`)
  * Constructor services support (service module plug-in)
  * Dependent constructor services support (service module plug-in)
  * Promisified service factories support (service module plug-in)
  * Registry adaptors (replaced by module plug-ins, where the CT service plug-in will need to be able to replace the BRJS service plug-in)
  * Service injection (`System.delete()` + `System.register()`)

**Comment by dchambers** [comment]
This was essentially fixed in abbad2b. This issue does however contain notes that continue to be relevant, and which will inform future SystemJS spike changes.
